### PR TITLE
Add Metro Bike Share Los Angeles, USA

### LINKED ISSUE #177: Add Los Angeles' Metro Bike Share
Official website: https://bikeshare.metro.net

It seems to be operated by BCycle, but the station map website is customised, so the existing scraper is unlikely to work. However, the station map is getting its data from this GeoJSON: https://bikeshare.metro.net/stations/json/

### PR DESCRIPTION
**Pull request by BenSto**
First Attempt to close  #177 . Feel free to improve or to even bring the scraper into the BCycle System, as mentioned in the Issue.

### CODE REVIEW

**Comment by BenSto** [comment]
Btw this scraper is already completely working.

File: pybikes/metro_bike_share.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    def update(self, scraper=None):
+        if scraper is None:
+            scraper = utils.PyBikesScraper()
+            response = scraper.session.request(
+            # some endpoints might fail verification, so it's up to the spider
+            # to disable it
+            url=self.feed_url,
+            method='GET',
+            verify=False,
+        )
```
Comment (eskerda) [inline pybikes/metro_bike_share.py]: This is not how this is supposed to work. We initialize an scraper in case we are not provided an scraper. Request happens on both cases

Comment (eskerda) [inline pybikes/metro_bike_share.py]: So:
```
scraper = scraper or utils.PyBikesScraper()
...
data = json.loads(scraper.request(self.feed_url))
```

File: pybikes/metro_bike_share.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+        data = json.loads(scraper.request(self.feed_url))
```
Comment (eskerda) [inline pybikes/metro_bike_share.py]: As it is now, it just makes two requests to same url, so I guess it's a typo

**Comment by eskerda** [comment]
`WIP Add Metro Bike Share...` is not a good commit name. Once it's added, it's no longer a WIP I guess?

File: pybikes/data/metro_bike_share.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{
+    "instances": [
+        {
+            "tag": "metro_bike_share",
```
Comment (eskerda) [inline pybikes/data/metro_bike_share.json]: Why do we use a tag that breaks with how other tags are named?

`metro-bike-share`

Comment (BenSto) [inline pybikes/data/metro_bike_share.json]: Sorry you are right, I was in rails convention, class name underscore.

File: pybikes/metro_bike_share.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+from .base import BikeShareSystem, BikeShareStation
+from . import utils
+
+class MetroBikeShare(BikeShareSystem):
+
+    sync = True
+
+    meta = {
+        'system': 'Metro Bike Share',
+        'company': 'Bicycle Transit Systems, Inc.',
```
Comment (eskerda) [inline pybikes/metro_bike_share.py]: I guess this should be

```
['Los Angeles County Metropolitan Transportation Authority (Metro)', 'Bicycle Transit Systems Inc', 'B-cycle']
```
or something similar

**Comment by BenSto** [comment]
> WIP Add Metro Bike Share... is not a good commit name. Once it's added, it's no longer a WIP I guess?

That was more to motivate people to transfer the scraper into the BCycle System. Because Im too lazy to do it. And in my opinion it doesnt make that much sense, cause the BCycle Scrapers have no logic in common with metro bike share.

File: pybikes/metro_bike_share.py (hunk 4 above)
Comment (BenSto) [inline pybikes/metro_bike_share.py]: Arrgghh another thing I forgot, that we can use arrays here. Sorry :-(

**Comment by eskerda** [comment]
> That was more to motivate people to transfer the scraper into the BCycle System. Because Im too lazy to do it. And in my opinion it doesnt make that much sense, cause the BCycle Scrapers have no logic in common with metro bike share.

You are right about that, it does not have any common logic, therefore they must sit on different codebase. Good choice

File: pybikes/metro_bike_share.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    def __init__(self, tag, feed_url, meta):
+        super(MetroBikeShare, self).__init__(tag, meta)
+        self.feed_url = feed_url
+
+    def update(self, scraper=None):
+        if scraper is None:
+            scraper = utils.PyBikesScraper()
+            scraper.ssl_verification = False
```
Comment (eskerda) [inline pybikes/metro_bike_share.py]: Why is this needed?

Comment (BenSto) [inline pybikes/metro_bike_share.py]: Otherwise I got that SSL Error:

```
  File "pybikes/metro_bike_share.py", line 25, in update
    data = json.loads(scraper.request(self.feed_url))
  File "pybikes/utils.py", line 65, in request
    verify=self.ssl_verification,
  File "/usr/local/lib/python2.7/dist-packages/requests-2.0.0-py2.7.egg/requests/sessions.py", line 357, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests-2.0.0-py2.7.egg/requests/sessions.py", line 460, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests-2.0.0-py2.7.egg/requests/adapters.py", line 358, in send
    raise SSLError(e)
requests.exceptions.SSLError: hostname 'bikeshare.metro.net' doesn't match either of '*.wpengine.com', 'wpengine.com'
```

**Comment by eskerda** [comment]
I've checked, and they seem to have banned default requests user agent. Pybikes uses another one, but this serves as an advisory that they do not like non-browser UA.

```
USER_AGENT = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.106 Safari/537.36'
...
...
scraper = scraper or utils.PyBikesScraper()
scraper.setUserAgent(USER_AGENT)
...
...

File: pybikes/metro_bike_share.py (hunk 2 above)
Comment (BenSto) [inline pybikes/metro_bike_share.py]: Yes that was an elder state, dont know why I commited that.

File: pybikes/metro_bike_share.py (hunk 5 above)
Comment (eskerda) [inline pybikes/metro_bike_share.py]: Weird, I can't reproduce.

Comment (BenSto) [inline pybikes/metro_bike_share.py]: > 

Weird, I can't reproduce.

>>> headers = {'User-Agent': 'Foobar'}
>>> requests.get('https://bikeshare.metro.net/stations/json/', headers=headers)
<Response [200]>
>>> 

Hmm thats really strange and I cant estimate, why you dont have that error too

Comment (eskerda) [inline pybikes/metro_bike_share.py]: full python version, OS?

Comment (BenSto) [inline pybikes/metro_bike_share.py]: Ubuntu 14.04 (Trusty Tahr)
```
 ~/pybikes   metrobike_los_angeles  python -V
Python 2.7.6
```

Comment (eskerda) [inline pybikes/metro_bike_share.py]: Maybe this is relevant then http://docs.python-requests.org/en/master/community/faq/#what-are-hostname-doesn-t-match-errors

Comment (BenSto) [inline pybikes/metro_bike_share.py]: Ok, thank you for the answer. I will try to upgrade my python version to 2.7.9. and see if the error disappears.

Comment (eskerda) [inline pybikes/metro_bike_share.py]: Since we are not checking on pybikes installation the python version, maybe it's a good idea to keep the verification off, with a comment on why it's like that

```
scraper = scraper or utils.PyBikesScraper()
scraper.setUserAgent(USER_AGENT)
# SSL verification fails on some python versions
# See: http://docs.python-requests.org/en/master/community/faq/#what-are-hostname-doesn-t-match-errors
scraper.ssl_verification = False
...
```

**Comment by jasongdove** [comment]
This data is publicly available via GBFS, as are most BCycle systems.

https://github.com/NABSA/gbfs
https://github.com/NABSA/gbfs/blob/master/systems.csv
https://gbfs.bcycle.com/bcycle_lametro/gbfs.json

**Comment by eskerda** [comment]
> This data is publicly available via GBFS, as are most BCycle systems.
> https://github.com/NABSA/gbfs
> https://github.com/NABSA/gbfs/blob/master/systems.csv

It does not appear on the CSV so I assumed somehow it wasn't

**Comment by eskerda** [comment]
https://github.com/eskerda/pybikes/issues/140

**Comment by jasongdove** [comment]
Yeah, I need to get that merged in the systems.csv - but you can also check the homepage https://bikeshare.metro.net for the GBFS link tag. Systems that support GBFS should have that tag.

<link rel="gbfs" type="application/json" href="https://gbfs.bcycle.com/bcycle_lametro/gbfs.json" />

File: pybikes/metro_bike_share.py (hunk 5 above)
Comment (BenSto) [inline pybikes/metro_bike_share.py]: > Since we are not checking on pybikes installation the python version, maybe it's a good idea to keep the verification off, with a comment on why it's like that

First sorry for the late answer. Guess you are right to leave a comment in the code as you don't pin the python version in this project. I'm going to add this like you already commented.

Comment (BenSto) [inline pybikes/metro_bike_share.py]: Ah Okay, you will use the GBFS Feed.

**Comment by eskerda** [comment]
https://github.com/eskerda/pybikes/pull/195 addes support for it. Going to close this PR for now. It's good we have it in case we needed this implementation for another system. Thanks @BenSto!
